### DISCUSSION TITLE
How can friction between different tools be minimized—especially with regard to rendering icons and diagrams?

### DISCUSSION THREAD

**Discussion by gwr-de** [OpenModelica/OpenModelica#12073] [Q&A] [answered]
Hi, I am the principal developer of the free [`Business Simulation Library (BSL)`](https://bsl-support.de/BusinessSimulation/BusinessSimulation) for Modelica, which I have just released for Wolfram System Modeler (`v2.2.0-wsm`). The library now declares dependency on Modelica Standard Library 4.0 and I would like to improve the experience for users of Open Modelica, so that I have decided to use a separate build for Open Modelica (and other tools), which will be based upon the version running in System Modeler.

For the following discussion I am referencing the current release-candidate (branch: `release`), which I have published in the library's public repository:

https://github.com/bslMS/BusinessSimulation/tree/release

**The good news** first: A lot of the examples already run without any issue in Open Modelica, which I checked with `v1.22.2 (64 bit)` on a laptop running `Windows 10` (I am developing on MacOs). That is good, as I am making use of the new unit framework in WSM 14 and had believed this might cause more issues in the examples.

**The bad news** are the looks. While I will expect deviations and difficulties from non-standard fonts etc., I cannot—as a developer—really live with the fact that lots of classes are not even close to what they look like in System Modeler? Some aren't even rendered at all. What are specs for then? :)

Let's look at a couple of diagrams: `Examples.SimpleProductionChainII` and `Examples.SimpleProductionChainIII` from the `Examples` package of the library.

|  Wolfram System Modeler 14 | Open Modelica 1.22.2  |
|---|---|
| ![`Examples.SimpleProductionChainII`](https://github.com/OpenModelica/OpenModelica/assets/17766670/91b2ebbf-e4bf-4a26-802c-d80b43a84b29) | ![`Examples.SimpleProductionChainII`](https://github.com/OpenModelica/OpenModelica/assets/17766670/ba37d2e8-48b7-44ed-ac0a-f593500403e9) |
| ![`Examples.SimpleProductionChainIII`](https://github.com/OpenModelica/OpenModelica/assets/17766670/ba390d53-786b-4570-a3d0-78ddcc42f0f2) | ![`Examples.SimpleProductionChainIII`](https://github.com/OpenModelica/OpenModelica/assets/17766670/037a3299-6efb-4a02-8e81-fb234a42dc34) |

You will have to admit that the diagrams in Open Modelica are pretty much unusable and will not instigate great appetite to try out the library?

Lots of "scripting error" messages suggest that `Smooth.Bezier` is unknown to Open Modelica and I believe other graphical annotations must be as well. So let's have a look at the very simple partial icon class `Icons.Info`.

|  Wolfram System Modeler 14  | Open Modelica 1.22.2  |
|---|---|
| ![`Icons.Info`](https://github.com/OpenModelica/OpenModelica/assets/17766670/14533d12-f4d1-4ce2-a59f-a539a2a1f508) | ![`Icons.Info`](https://github.com/OpenModelica/OpenModelica/assets/17766670/7658fb8d-3347-4ca8-b37b-7b2117c16c15) |

That icon is not rendered in Open Modelica _at all_. Here is the code:

```
partial class Info "Icon for information pages (UsersGuide)"
  annotation(Icon(coordinateSystem(extent = {{-100, -100}, {100, 100}}, preserveAspectRatio = true, initialScale = 0.1, grid = {10, 10}), graphics = {Ellipse(visible = true, lineColor = {255, 170, 0}, fillColor = {255, 170, 0}, fillPattern = FillPattern.Solid, lineThickness = 5, extent = {{-95, -95}, {95, 95}}), Polygon(visible = true, origin = {-4.167, -15}, fillColor = {255, 255, 255}, pattern = LinePattern.None, fillPattern = FillPattern.Solid, points = {{-15.833, 20}, {-15.833, 30}, {14.167, 40}, {24.167, 20}, {4.167, -30}, {14.167, -30}, {24.167, -30}, {24.167, -40}, {-5.833, -50}, {-15.833, -30}, {4.167, 20}, {-5.833, 20}}, smooth = Smooth.Bezier), Ellipse(visible = true, origin = {7.5, 56.5}, fillColor = {255, 255, 255}, pattern = LinePattern.None, fillPattern = FillPattern.Solid, extent = {{-12.5, -12.5}, {12.5, 12.5}})}), Diagram(coordinateSystem(extent = {{-150, -90}, {150, 90}}, preserveAspectRatio = true, initialScale = 0.1, grid = {5, 5})));
end Info;
```

Can you, please, provide some guidance for us poor developers, who hope that _"write once, run everywhere"_ is a truism for Modelica? :)

**Comment by casella** [comment]
@gwr69 thanks for you feedback! I am a true believer in Modelica libraries that run in multiple tools (I am involved in a number of them), so I'm willing to help you achieving this goal.

The first problem here is that many (most?) users of OpenModelica use OMC and the Modelica Standard Library to build their stuff, so by definition they get what they want. We need more users like you that are developing on tool X and then want to also run their libraries in OMC.

Second, as soon as we get our systems online again (today there were problems on the network were our machines are hosted) I will add your branch https://github.com/bslMS/BusinessSimulation/tree/release to our [library testsuite](https://libraries.openmodelica.org/branches/overview.html), so we can track its performance.

Third, there are probably some bugs we need to fix. For that, the rule is: one bug, one issue on the issue tracker. That's the only way we can manage them. Preferably with a minimal working example that allows to reproduce the issue. Please add them without restraint 😃. As we are heavily improving the GUI (see the [list of issues](https://github.com/OpenModelica/OpenModelica/issues?q=label%3ACOMP%2FGUI%2FOMEdit+is%3Aissue+closed%3A2023-03-05..2024-03-05+is%3Aclosed) fixed in the last 12 months) I would suggest you to try with the latest nightly build before opening a ticket, as the issue may have already been solved. I already tried myself, and it seems the issues you report here are sill valid.

Once we have your library in our Package Manager I'll start opening tickets about the issues you raised here. I am pretty sure they will be solved quickly. Then, we can see how we go from there.

BTW, I wrote a paper (a book chapter, actually, http://dx.doi.org/10.1007/3-7908-1636-1_30) about modelling the bullwhip effect with Modelica, together with my colleague Giovanni Miragliotta of Politecnico di Milano and his PhD student Luigi Uglietti, almost 20 years ago. For various reasons we did not continue that work, but we're still in contact, I think Giovanni may be interested in your work.

**Reply by gwr-de** [reply]
@casella Great to feel welcome here and the "fixing of issues" is already going strong (#12077, #12074)!

With regard to accommodating the BusinessSimulation library I see these areas:

### Graphical Display

I believe this to look a lot worse than it actually is and I am very hopeful that the display of diagrams and icons (hopefully also in Libraries Browser...) is going to look nice pretty soon. Unfortunately, there is still no real agreement within the language group as to what is definitely determining connection line styling and to what extend these connections should be open to determination by a library ([#3380](https://github.com/modelica/ModelicaSpecification/issues/3380)). :)

### Connection Semantics and Convenience

Unlike more physical components, the number of connectors (and parameters) in the realms of business and economics tend to grow quite fast—supply chain modeling is a good case in point. So far, I find that neither `composite connectors` nor `expandable connectors` are a good fit for pre-built components, given the restrictive connection semantics? Currently, `stacked connectors` as supported by Dymola would be a simple fix, if the GUI supports connections with a nice drop-down menu. Not sure, whether OMEdit will go along here?

### Custom Derived and Base Unit Definitions

The convenience offered by WSM 14 with regard to derived and base unit definitions and the rather nice unit string parsing to determine `displayUnit` conversions from these, may serve as a good example to ponder. I am not sure, whether OM will move along here all the way but I believe at least a bit more flexibility is needed. Not only in the "soft sciences" will modelers want to specify simulation time spans and process durations in `years` (e.g., Gregorian average years or tropical years or Julian years)—not seconds.

As I had written in the [Mole Day Proposal](https://github.com/modelica/ModelicaSpecification/issues/3425) to the language group, something is wrong, if even experts in the field (SystemDynamics library) use seconds while they mean years, in fact. Maybe you guys can point me to what can be done in Open Modelica currently as to define custom conversion and custom units in libraries?

**Looking forward very much to a nice experience for BusinessSimulation library users in OMEdit coming soon!**

**Reply by casella** [reply]
I'm not aware of stacked connectors. Where can I find information about them?

**Reply by gwr-de** [reply]
I learned it [here](https://github.com/modelica/ModelicaSpecification/issues/3206).

**Reply by casella** [reply]
Just for my understanding, what would be the advantage of those stacked connectors vs. connectors that contain all the sub-connectors (or simply the variables) that correspond to the stacked connectors?

**Reply by gwr-de** [reply]
My understanding is that `composite connectors` (`cc`) are intended like a single plug that connects `a.cc` with `b.cc` in a single connection. This is great in contexts where we know all the variables that will be connected between `a` and `b`, e.g.,  the material, transactional connections bewteen a suplier and a producer.

But this is often not true for control and information inputs where we we have lots of input eventually from different sources. Then Modelica’s restiction that we can’t connect `a.y1` with `b.cc.u1` necessitate that we need a copy of `cc` at the models’ scope as an intermediary “switch.” This becomes even more unwieldy when there are `input` and `output` designations.

The `expandable connector` is nicer in handling, which is why I am using it for more complex components like `MoleculesOfStructure.Acuators.Diffusion` (see `Examples.SimpleProductionChainIII`) and `Flows.Interaction.ComplexInteraction` (see `Flows.Interaction.LotkaVolterra` for how this need to be connected to by a wrapper class).

But even the expandable connector does not directly allow to connect `a.y1` with `b.ec.u1`, does it? For a pre-built class that is what we want: Nice and simple connection, but not having a component icon that looks like a “porcupine.”

What have I overlooked? What other modeling option will give “simple convenience” where lots of differently labeled (which is why `array` is not good) need to be connected to different other elements?

**Reply by casella** [reply]
Good question.

To me, Expandable connectors are a metaphor of bus connections on a DCS. You have a hierarchically structured plant model, and a hierarchically structured controller model. Signals are carried around on connected Bus connectors via proper labelling, e.g., somewhere in the plant a temperature sensor will output `Temperature0021` on the bus connector, and somewhere in the controller this signal will be picked up by some temperature PID controller by its name. Without the need of spaghetti diagrams to connect each output to each input, you just connect buses together and they move around whatever signals they may carry. But I always though about expandable connectors as flat connectors. Of course you can structure them hierarchically, putting together signals of different categories, but it's just for tidiness reasons. As I understand, that's not what you are looking for.

If I got it right, you want to be able to have bunches of signal, say, `a, b, c, d`, whereby you can connect all of them from a source output to a sink input with just one connection line, or possibly connect sub-bunches, say you get the inputs for `a,b` from one source and from `c,d` from another source, with two connection lines. You could already do that with array connectors using sub-ranges, à la Simulink, but that would be ugly. Are we on the same page?

I would suggest you to open another discussion on this topic and move it there, otherwise this discussion item will soon be unmanageable :)

**Reply by gwr-de** [reply]
Done: #12088

**Comment by casella** [comment]
In the meantime, I tried to open your Production Chain II into Dymola 2024X. This is what I got:
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/fd3922fc-6a8e-4ed9-9031-68d04ee33ca5)
Currently better than what you got with OMEdit, but still notably different from what you get in SystemModeller.

Apart from Dymola having different opinions about connection routing, I see that the line thickness, though not as phat as in the case of OMEdit, is still notably thicker that what you get in SystemModeller. I guess we need a specific test case for that. I'll take care of it ASAP.

**Reply by gwr-de** [reply]
The letter “R” in the Icon for the inherited `Type.Reals` is visible for the causal connectors throughout. That is ugly.

**Comment by casella** [comment]
The fat lines issue is addressed in #12090. I think I had similar problems in the past, so it's good that you raised it.

**Comment by casella** [comment, marked as answer]
Once #12090 is also fixed, graphical rendering should be in good shape 😃

**Reply by gwr-de** [reply]
That sounds great and I harbor the realistic hope that I will then just need slight adaptations to make everything look as nice as it should. Then we can focus on display units for time and be (almost) done. 😀

**Reply by gwr-de** [reply]
Since #12090 is closed now, the graphical disparities should have vanished or rendered "minor variations" by now.\
Can you point me as to whether in OMEdit there is a way to define derived units of time to use with `displayUnit`, e.g., conversion factors for something like [`yrCal`](https://bsl-support.de/BusinessSimulation/BusinessSimulation.Units.Time_years.html)?

**Reply by gwr-de** [reply]
As a clarification: The issue arises especially for the internal variable `time` in Modelica, which cannot be changed to any other unit of time than _seconds_ as far as I know. In business, economics, ecology, and even STEM-related models a tool might support the convenience to provide experiment settings (e.g. `StartTime`, `StopTime`) using `displayUnit` comparable to what WSM does in the model annotations for `BusinessSimulation.Examples.SimpleProductionChain`):

```
model SimpleProductionChain
  ...
  annotation(experiment(StartTime = 0, StopTime = 157784760, __Wolfram_DisplayTimeUnit = "mo"), ...)
end SimpleProductionChain;
```

**Reply by casella** [reply]
Unfortunately non-SI units are not defined by the Modelica Language Specification (which is a problem IMHO), so they are pretty much tool dependent. We have thought with @adrpo and @adeas31 of making this definition easily accessible by end users for personalization, but then this won't be portable, and that's unfortunately a language issue, not a tool issue.

We could start a discussion about that on the MLS issue tracker, but there is already a heated and long debate on units, so I'm afraid it won't end anywhere. Maybe a practical solution which allows you to customize OMEdits non-SI units would be good enough for the time being. 

@adeas31, @adpro, what do you think?

BTW, I had this very same issue when developing simple models for COVID-19 dynamics, where of course the unit of time should be days, not seconds. Eventually I ended up pretending seconds are days, see https://github.com/casella/COVID19Control/tree/master, but that's not nice at all.

**Reply by gwr-de** [reply]
Indeed, watching MLS discussions play out is not a “spectator sport” … :)

I could well live with a customized OMEdit version and to be mighty honest, I feel that having BSL-related things run smoothly in WSM and OMEdit will likely service at least 95% of the intended target group. I see Dymola as an “engineering only” tool, which is fine, but see OMC and WSM/WL addressing broader interests.

**Reply by gwr-de** [reply]
Just wanted to give a quick feedback here: The graphics now look really nice after I managed to hide the type-related icon layers with the type letter "R" appearing behind the output connectors. Great work everyone!

**Reply by casella** [reply]
Good to hear!. If there are any remaining issues, please open tickets about them.

**Reply by gwr-de** [reply]
@casella Since graphical issues are solved as far as I can see, I would like to proceed with my release. Does it make sense to wait for some preliminary decision (framework) on how library developers could store unit conversion factors for OMEdit as you proposed above? From my point of view, this would predominantly concern units of `time` and dimensionless rates. In System Modeler they had vendor annotations for this:

```
__Wolfram(unitConversions(conversion(derivedUnit = "wk", unit = "s", factor = 604800., offset = 0., description = "Time"), ...
```

Maybe something in this vein is possible in OMEdit as to not be dependent on "endless discussions" in the language group?

**Reply by casella** [reply]
@gwr69 I would just try to add your library to the library testsuite, to see if there are serious issues running some of the models you have there. That was in my TODO list last week but I got drowned and I'm emerging right now. I can take care of that today or tomorrow. I would wait for that, so if there are more fixes you can do them right away.

> In System Modeler they had vendor annotations for this:
```
__Wolfram(unitConversions(conversion(derivedUnit = "wk", unit = "s", factor = 604800., offset = 0., description = "Time"), ...
```
This seems very reasonable to me, I guess we could implement it as an __OpenModelica custom annotation without too much fuss.

@adeas31, @adrpo, what do you think?

**Reply by gwr-de** [reply]
@casella That sounds great. My original conversion annotations are still visible on the `main` branch of my repo:

https://github.com/bslMS/BusinessSimulation/blob/0130f5e67d46e3cdbcf7ef2e5d7af9078d805ba8/BusinessSimulation/package.mo

I wanted to adapt the examples as they are making use of units that at the moment are not directly supported, e.g., `each`. Would you deem this necessary/helpful?

**Reply by casella** [reply]
the use of `each` when applying scalar modifiers to arrays is compulsory, according to the Modelica Specification, despite what some tool do. I understand trying to figure that out automatically is non-trivial, so we're not going to invest time on that. Please make sure `each` is used in all places where it should.

**Reply by gwr-de** [reply]
Completely d'accord. Can you point me to an omission? System Modeler typically raises errors, so I am rather confident that we are talking about at most a handful of cases ...

**Reply by casella** [reply]
I just started the process to add the version of `BusinessSimulation` on the `main` branch to our library testsuite. The report will be marked `BusinessSimulation_dev`, you'll find it [here](https://libraries.openmodelica.org/branches/master/) in one-two days, please check that for any remaining non-compliance issue or OMC bug. Note that we automatically test models that have an `experiment(StopTime)` annotation, please make sure that all models you want to test (and only them, not, e.g., component models in the library) have it.

Once you have a release that is known to work well, I can add one more test to the release version, so we can tell apart regressions due to library development and regressions caused by bugs in OMC. If you want to provide reference results as .mat or .csv file, we can also check the results against them.

In fact, it turns out @sjoelund added your library to the package manager four years ago as "experimental". But we never tested it. Once the tests are (mostly) fine, we can mark it as "support" or "fullSupport" in the package manager.

**Reply by gwr-de** [reply]
@casella The version on the main branch is not compatible with MSL 4.0.0, which is why the release of `v2.2.0` is rather important now. I have validated all classes in Wolfram System Modeler and I am employing a battery of more than 300 tests using `VerificationTest` in WolframLanguage and the tight integration of System Modeler with it. I would likely need to take recourse to FMI calls, which to my knowledge are not support in WL as of now?

**Reply by casella** [reply]
> I have validated all classes in Wolfram System Modeler and I am employing a battery of more than 300 tests using VerificationTest in WolframLanguage and the tight integration of System Modeler with it. I would likely need to take recourse to FMI calls, which to my knowledge are not support in WL as of now?

I'm a bit lost here. In most cases, example in libraries work out of the box, possibly needing to read some files from the Resource directory. This can be handled entirely via annotations in the library. In some cases, people have test scripts that need to do additional stuff before running the simulation. Can you describe in a bit more detail what do you need, maybe in a separate ticket?

**Reply by gwr-de** [reply]
@casella 

A lot of my tests are building the model computationally (i.e., by using [`ConnectSystemModelComponents`](https://reference.wolfram.com/language/ref/ConnectSystemModelComponents.html)) and then just run a script that checks for given results ([VerificationTest](https://reference.wolfram.com/language/ref/VerificationTest.html?q=VerificationTest)). I see that there is scipting in OM so I could more or less easily build a package of example models that can be run by a script and then checked against given results? While I am learning Julia, Wolfram Language is my go to choice - I am writing it "in my sleep"(almost) ... :)

**Reply by casella** [reply]
@gwr69 testing libraries with scripts is of course a sensible option, and many people do that. Unfortunately those scripts are not standardized, so if you have a Modelica library that you want to run on multiple tools, you need to write them for each tool. OpenModelica has an [API](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html) that would allow you to do something similar to what you describe, but you'd need to port your current SystemModeler scripts to it manually. 

If want to do that, we can discuss ways to incorporate it in our testing infrastructure. A simple option could be that you provide .mos scripts from some git repo, we just fetch them and run them. In fact, we already do this for the core testsuite that is run on every pull request. The problem is how to represent the outcome of those tests - for our core testsuite, the output is a bit ugly, basically something like a textual diff file, it's not as nice as the [test reports](https://libraries.openmodelica.org/branches/master/BusinessSimulation_dev/BusinessSimulation_dev.html) you get from running models that are defined in the library with an experiment annotation.

If this is interesting for you, please open another discussion item about it so we keep the whole thing manageable, as this discussion is becoming a tad too crowded...

**Comment by adeas31** [comment]
Was just looking into the rendering differences and saw that `BusinessSimulation.Converters.ConstantConverter` is a mess. It contains several `Icon` and `Diagram` annotations. Probably the tool that you used to develop the library added them.

So when rendering `BusinessSimulation.Converters.ConstantConverter` icon, we are considering the last annotation and apparently Dymola and WSM are using the first one. @perost should we issue a warning in this case?

**Reply by perost** [reply]
Having multiple Icon/Diagram annotations is as far as I can see undefined, the specification doesn't give any restrictions on how many you can have and doesn't say what happens if there's more than one such annotation. So we can do whatever we want in that case, including giving a warning.

**Reply by gwr-de** [reply]
Good catch! That's likely some old copy and past error that got forgotten. I'll correct this on the `release` branch. Thanks.

**Reply by gwr-de** [reply]
A tool imho should definitely raise a warning as this is very hard to spot.

**Reply by gwr-de** [reply]
I removed hopefully these cases (with my last commit), which amounted to `ConstantConverter` related classes (4) only. Using RegEx I was not able to spot any other occurrences—but maybe I should set up a Bounty for finding one? :)

**Comment by gwr-de** [comment]
I just note the following related "issues" and wonder whether they in fact are _issues_:

### Zooming will not affect text without a fixed font size in the same way as other graphical elements

The library defines `type Reals` in its `Types` package and provides (as is often custom) an Icon that essentially consists of the letter `"R"` which is used with coordinate-related scaling—not a fixed font size. This class is a base class for `connector RealOutput`, which is defined in the `Interfaces.Basics.OutputTypeChoice` class using a `replaceable type OutputType`. In line with specs the inherited graphical elements are put in the back and the "triangle" for the connector is put in front.

In Wolfram System Modeler 14 the letter is completely _hidden behind the triangle_ in the  Icon View and this is also the case for Open Modelica in today's nightly build. But in diagrams, such as `Examples.SimpleProductionChainII` the visibility of the letter `"R"` depends on the zoom factor chosen, i.e., zooming out will reveal the letter—zooming in will hide it again. Since I have defined the letter without a fixed font size, it should be scaled in the same way all other graphical elements are and thus if it is invisible in Icon View that should not change due to zooming, which is the behavior I am experiencing in System Modeler.

### Support for IconMap() and DiagramMap() annotation

To avoid this issue, one might simply avoid graphical elements from being inherited from a base class. But adding the annotation `IconMap(primitivesVisible = false)` to either `replaceable type OutputType = ...` or to `connector RealOutput = ...` does not seem to have any effect. Is it supported?

**Reply by adeas31** [reply]
> I just note the following related "issues" and wonder whether they in fact are _issues_:
> 
> ### Zooming will not affect text without a fixed font size in the same way as other graphical elements
> The library defines `type Reals` in its `Types` package and provides (as is often custom) an Icon that essentially consists of the letter `"R"` which is used with coordinate-related scaling—not a fixed font size. This class is a base class for `connector RealOutput`, which is defined in the `Interfaces.Basics.OutputTypeChoice` class using a `replaceable type OutputType`. In line with specs the inherited graphical elements are put in the back and the "triangle" for the connector is put in front.
> 
> In Wolfram System Modeler 14 the letter is completely _hidden behind the triangle_ in the Icon View and this is also the case for Open Modelica in today's nightly build. But in diagrams, such as `Examples.SimpleProductionChainII` the visibility of the letter `"R"` depends on the zoom factor chosen, i.e., zooming out will reveal the letter—zooming in will hide it again. Since I have defined the letter without a fixed font size, it should be scaled in the same way all other graphical elements are and thus if it is invisible in Icon View that should not change due to zooming, which is the behavior I am experiencing in System Modeler.

This is probably because how we draw the text annotation. We use the flag [Qt::TextDontClip](https://doc.qt.io/qt-6/qt.html#TextFlag-enum). This makes it possible to render the text even if the bounding box of it is very small. I noticed the same behaviour in Dymola.

| OpenModelica | Dymola |
| ---| --- |
| ![image](https://github.com/OpenModelica/OpenModelica/assets/4007231/13d44d1c-33d7-4717-9bb8-2f952cf20c49) | ![image](https://github.com/OpenModelica/OpenModelica/assets/4007231/4ee47fe6-a766-431f-9699-5989d0a0921d) |

> 
> ### Support for IconMap() and DiagramMap() annotation
> To avoid this issue, one might simply avoid graphical elements from being inherited from a base class. But adding the annotation `IconMap(primitivesVisible = false)` to either `replaceable type OutputType = ...` or to `connector RealOutput = ...` does not seem to have any effect. Is it supported?

`IconMap` and `DiagramMap` annotations are supported. If you are having issues with it then please open a new ticket about it with minimal example model.

**Comment by AustraleanPerson** [comment]
> This makes it possible to render the text even if the bounding box of it is very small. I noticed the same behaviour in Dymola.

There is a very useful flag "Restrict minimum font size" in Dymola (default ON). It might be worth adding a similar feature in OM.

**Reply by casella** [reply]
It's useful, but also leads to poorly portable code, because then people rely on it (possibly without even being aware of it) and produce diagrams which then will look very bad on other tools. Unless they implement the de-facto standard instead of the real one, i.e. the Modelica Language Specification.

Food for thought... 😃

**Reply by gwr-de** [reply]
Exactly, but to be fair: I remember a discussion with a person appearing in the list of contributors to your project, who more or less explicitly hinted at there being in fact "reference implementations" out there—while "naive me" had believed Modelica would work with reference specification instead.

Food for self-reflexion? 😉

**Reply by gwr-de** [reply]
Now, that it is fixed, I can also reveal that at the time I had complained (already) about the poor quality of graphic rendering compared to System Modeler. I was decidedly told "Well, it is your choice whether to go for the tools most people use or ..."

**Reply by casella** [reply]
It's a long story. My point is, it is a fact of life that we are not standardizing everything at once, but if there are "reference implementations" filling voids in the specification, we should strive to complete the specification.

**Comment by casella** [comment]
> Now, that it is fixed, I can also reveal that at the time I had complained (already) about the poor quality of graphic rendering compared to System Modeler. I was decidedly told "Well, it is your choice whether to go for the tools most people use or ..."

It is our goal to have a full-fledged Modelica environment, not a poor man's one. Of course we have to deal with limited resources. But it is important to be aware that OpenModelica's limitations are not etched in stone :)

BTW, do you work for a company that would be interested at joining the Open Source Modelica Consortium?

**Reply by gwr-de** [reply]
> BTW, do you work for a company that would be interested at joining the Open Source Modelica Consortium?

I am running my own company, which also means that I have to think carefully about where I spent my resources. I need to turn part of the BusinessSimulation efforts into making some money from it...

**Reply by casella** [reply]
OK. The consortium fee for a very small company (<= 5 employees) is 1200 €/yr. If you start getting revenue from using OMC, please consider that.

**Reply by gwr-de** [reply]
Yes, I will definitely consider it.

**Comment by gwr-de** [comment]
@casella I have now adapted all the examples to run in `s` (stretching this to mean "standard units of time for the model"). This is not optimal, but should likely give best user experience as long as a nice way to have "months" or "years" is missing.

A thing that I find annoying is that `annotation(Dialog(group = "Structural Parameters"));` seems to be ignored?  In System Modeler there is a tab for parameters, which will have different groups. I also cannot see any "tabs" for things like `annotation(Dialog(tab = "Initialization"));`?

Is there a way to change parameters _before_ a simulation is started in a "nice" way? So far, I see ways to change parameters _after_ a simulation has been simulated in the "variables" windows, but no way to do this before a simulation has started. This is the more important for "Structural Parameters" (which in my library always go into `group = "Structural Parameters"`). These parameters will typically get `Evaluate = true`, which unfortunately prevents a user from switching on a conditional component in say the `HealTheWorld` example.

What can I do and what can you do in this regard?

**Reply by adeas31** [reply]
> A thing that I find annoying is that annotation(Dialog(group = "Structural Parameters")); seems to be ignored? In System Modeler there is a tab for parameters, which will have different groups. I also cannot see any "tabs" for things like annotation(Dialog(tab = "Initialization"));?

Something that doesn't have a tab defined should end up in "General" tab AFAIR.

**Reply by gwr-de** [reply]
But it doesn't for me - there is no "General" tab, which means that while I can get a parameters menu option for some component upon a right click, there is no such thing for the model itself?

**Reply by adeas31** [reply]
We do not yet support editing of top level parameters of a model. I see that you created #12184 for it.

**Reply by gwr-de** [reply]
Yes, as this omission leads to a rather bad user experience for my library, currently. It now runs fine and I adapted all the settings to "seconds" (for now), but not being able to change top level parameters easily renders most of my examples "painful" ...

**Reply by casella** [reply]
See #4408
